packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -0,0 +1,135 @@
+// proposals, which can be disputed and settled.
+const { OptimisticOracleRequestStatesEnum, averageBlockTimeSeconds } = require("@uma/common");
+const Promise = require("bluebird");
+
+class OptimisticOracleClient {
+  /**
+   * @notice Constructs new OptimisticOracleClient.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} oracleAbi OptimisticOracle truffle ABI object to create a contract instance.
+   * @param {Object} web3 Provider from Truffle instance to connect to Ethereum network.
+   * @param {String} oracleAddress Ethereum address of the OptimisticOracle contract deployed on the current network.
+   * @param {Number} lookback Any requests, proposals, or disputes that occurred prior to this timestamp will be ignored.
+   * Used to limit the web3 requests made by this client.
+   * @return None or throws an Error.
+   */
+  constructor(
+    logger,
+    oracleAbi,
+    web3,
+    oracleAddress,
+    lookback = 604800 // 1 Week
+  ) {
+    this.logger = logger;
+    this.web3 = web3;
+
+    // Oracle contract
+    this.oracle = new web3.eth.Contract(oracleAbi, oracleAddress);
+    this.oracleAddress = oracleAddress;
+
+    // Oracle Data structures & values to enable synchronous returns of the state seen by the client.
+    this.priceRequests = [];
+    this.priceProposals = [];
+
+    // Store the last on-chain time the clients were updated to inform price request information.
+    this.lastUpdateTimestamp = 0;
+    this.lookback = lookback;
+
+    // Helper functions from web3.
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+  }
+
+  // Returns an array of Price Request objects that have no proposals yet.
+  getAllPriceRequests() {
+    return this.priceRequests;
+  }
+
+  // Returns an array of Price Request objects that someone has proposed a price for.
+  getAllPriceProposals() {
+    return this.priceProposals;
+  }
+
+  // Returns an array of Price Request objects for each position that someone has proposed a price for and whose
+  // proposal can be disputed. The proposal can be disputed because the proposed price deviates from the `inputPrice` by
+  // more than the `errorThreshold`.
+  getDisputablePriceProposals() {
+    // TODO
+    return;
+  }
+
+  // Returns the last update timestamp.
+  getLastUpdateTime() {
+    return this.lastUpdateTimestamp;
+  }
+
+  async update() {
+    // Determine earliest block to query events for based on lookback window:
+    const [averageBlockTime, currentBlock] = await Promise.all([
+      averageBlockTimeSeconds(),
+      this.web3.eth.getBlock("latest")
+    ]);
+    const lookbackBlocks = Math.ceil(this.lookback / averageBlockTime);
+    const earliestBlockToQuery = Math.max(currentBlock.number - lookbackBlocks, 0);
+
+    // Fetch contract state variables in parallel.
+    const [requestEvents, currentTime] = await Promise.all([
+      this.oracle.getPastEvents("RequestPrice", { fromBlock: earliestBlockToQuery }),
```

Comment:
Starting a new thread so this is more clear, I think we should consider any requests that have been requested, proposed, or disputed in this time range, since you could have a request that has no reward and therefore sits without a proposal for a long time. Once someone decides to propose a price for that request, we should consider it for disputes, etc. Thoughts?

Reply:
I agree, and I was going to iterate on this on the next PR. Here's the problem how I see it:
- Every unique Proposal and Dispute event is also associated with a RequestPrice

So the way I was going to query for Proposals & Disputes was: query for all RequestPrices then call `getState()` on each one to sort it.

On the other hand, we could run event queries for Disputes and Proposals. I'm not super sure yet which strategy results in fewer web3 calls but was going to tackle this in the next issue.

Reply:
Tacking in the next PR SGTM. The issue I see with the way it's being done now is just that a request can sit for weeks (if it's unrewarded, for instance) and then get a proposal that requires input from our bot (or else could return a bad price). This is really only the case if we force our bot to only make proposals for requests that are rewarded. Otherwise, we might expect that our bot ensure no requests sit for very long.